crates/sqlite-lambda/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+            field: field.to_string(),
+            ptr: doc::Pointer::from_str(ptr),
+            is_format_integer,
+            is_format_number,
+            is_content_encoding_base64,
+        }
+    }
+}
+
+/// Lambda is a SQLite execution context which is able to repeatedly invoke a
+/// given query over novel source and register documents, returning an array
+/// of document results.
+pub struct Lambda<N: doc::AsNode + 'static> {
+    db: Option<rusqlite::Connection>,
+    stmt: Option<rusqlite::Statement<'static>>,
+    output_columns: Vec<String>,
+
+    /// Cells which hold documents which will be taken by the next vtab::Cursor.
+    source: Rc<Cell<Option<&'static N>>>,
+    register: Rc<Cell<Option<&'static N>>>,
+    previous_register: Rc<Cell<Option<&'static N>>>,
+}
+
+impl<N: doc::AsNode + 'static> Lambda<N> {
+    /// Create a new Lambda which executes the given `query`. The execution
+    /// context will include a table `source` having the given `source_columns`.
+    ///
+    /// If `register_columns` is non-empty, it will additionally have tables
+    /// `register` and `previous_register` having the given `register_columns`.
+    ///
+    /// Each row returned by `query` is mapped into a JSON document, with each
+    /// output column becoming a top-level document property. SQLite types are
+    /// mapped to corresponding JSON types null, integer, float, and string.
+    ///
+    /// Nested JSON arrays and objects are also supported: As SQLite doesn't have
+    /// a bespoke JSON value type, this implementation looks for a leading / trailing
+    /// pair of '{','}' or '[',']' and, if found, will attempt to parse the string
+    /// as a JSON document. If parsing fails, the raw text is passed through as a
+    /// regular JSON string.
+    ///
+    /// As a special case, if the query has a single output column named `flow_document`
+    /// then this column is directly mapped into the returned output document.
+    /// This can be used to implement lambdas with dynamic top-level properties.
+    pub fn new(
+        query: &str,
+        source_columns: &[Projection],
+        register_columns: &[Projection],
+    ) -> rusqlite::Result<Self> {
+        let db = rusqlite::Connection::open_in_memory()?;
+
+        let source = vtab::Table::<N>::install(&db, "source", source_columns)?;
+        let (register, previous_register) = if !register_columns.is_empty() {
+            (
+                vtab::Table::<N>::install(&db, "register", register_columns)?,
+                vtab::Table::<N>::install(&db, "previous_register", register_columns)?,
+            )
+        } else {
+            Default::default()
+        };
+
+        let stmt = db.prepare(query)?;
+
+        // Extract output columns from the prepared statement.
+        let output_columns = if stmt.column_name(0)? == "flow_document" && stmt.column_count() == 1
+        {
+            Vec::new()
+        } else {
+            (0..stmt.column_count())
```

Comment:
:shrug: it returns &str and we need to avoid having a borrow of `stmt` so we can call `stmt.query(&mut self)`.

Either way it's going to require mapping into strings and re-collecting into a Vec.